synapse/handlers/account_validity.py

```diff
@@ -223,37 +223,70 @@ async def _get_renewal_token(self, user_id: str) -> str:
                 attempts += 1
         raise StoreError(500, "Couldn't generate a unique string as refresh string.")
 
-    async def renew_account(self, renewal_token: str) -> bool:
+    async def renew_account(self, renewal_token: str) -> Tuple[bool, bool, int]:
         """Renews the account attached to a given renewal token by pushing back the
         expiration date by the current validity period in the server's configuration.
 
+        If it turns out that the token is valid but has already been used, then the
+        token is considered stale. Stale status is calculated by checking if an account
+        associated with the token has the `email_sent` DB column set to False. When a
+        renewal token if first sent out, this column is True. When the user renews using
+        this token, the column is set to False.
+
         Args:
             renewal_token: Token sent with the renewal request.
         Returns:
-            Whether the provided token is valid.
+            A tuple containing:
+              * A bool representing whether the token is valid.
+              * A bool representing whether the token is stale.
+              * An int representing the user's expiry timestamp as milliseconds since the
+                epoch, or 0 if the token was invalid.
         """
         try:
-            user_id = await self.store.get_user_from_renewal_token(renewal_token)
+            (
+                user_id,
+                email_sent,
+                current_expiration_ts,
+            ) = await self.store.get_user_from_renewal_token(renewal_token)
         except StoreError:
-            return False
+            return False, False, 0
+
+        # True if a user is found and email_sent is False, meaning the token has already
+        # been used. False otherwise.
+        if not email_sent:
+            logger.warning(
+                "User %s attempted to use previously used token to renew account",
+                user_id,
+            )
```

Comment:
Is this really a warning? Are we considering this potentially malicious?

It is probably worth including the token in the log line?

Reply:
I considered it useful to log in case a user asks why their renewal date isn't updating. But this information could be gleamed from checking what template they were receiving + looking at request logs and their `account_validity` table entry.

If you're asking about warning versus info, yeah perhaps a *warning* doesn't quite make sense here.

Reply:
I was more asking about warning vs. info. Logging something is reasonable. 😄

Reply:
Gotcha. Changed to `info` and am now logging the token.